kart/sqlalchemy_import_source.py

```diff
@@ -0,0 +1,242 @@
+import functools
+import os
+import sys
+
+import click
+import sqlalchemy
+
+from .exceptions import (
+    NotFound,
+    NotYetImplemented,
+    NO_TABLE,
+)
+from .geometry import Geometry
+from .import_source import ImportSource
+from .sqlalchemy import DbType, separate_last_path_part
+from .output_util import dump_json_output
+from .utils import ungenerator, chunk
+
+
+class SqlAlchemyImportSource(ImportSource):
+    """
+    ImportSource that uses SqlAlchemy directly to import into Kart.
+    Currently only GPKG is supported, but in theory should work for Postgis, SQL Server, MySQL.
+    """
+
+    CURSOR_SIZE = 10000
+
+    @classmethod
+    def open(cls, spec):
+        db_type = DbType.from_spec(spec)
+        if db_type is None:
+            raise cls._bad_import_source_spec(spec)
+
+        # TODO - add support for other DB types.
+        if db_type is not DbType.GPKG:
+            raise NotYetImplemented(
+                "Only GPKG is currently supported by the SqlAlchemyImportSource"
+            )
+
+        path_length = db_type.path_length(spec)
+        longest_allowed_path_length = db_type.path_length_for_table
+        shortest_allowed_path_length = max(
+            db_type.path_length_for_table_container - 1, 0
+        )
+
+        if (
+            path_length > longest_allowed_path_length
+            or path_length < shortest_allowed_path_length
+        ):
+            raise cls._bad_import_source_spec(spec)
+
+        connect_url = spec
+        db_schema = None
+        table = None
+
+        # Handle the case where specification already points to a single table.
+        if path_length == db_type.path_length_for_table:
+            connect_url, table = separate_last_path_part(connect_url)
+            path_length -= 1
+
+        # Handle the case where specification points to a database schema (or similar).
+        if path_length > shortest_allowed_path_length:
+            connect_url, db_schema = separate_last_path_part(connect_url)
+
+        engine = db_type.class_.create_engine(connect_url)
+        return SqlAlchemyImportSource(spec, db_type, engine, db_schema, table)
+
+    @classmethod
+    def _bad_import_source_spec(self, spec):
+        return click.UsageError(
+            f"Unrecognised import-source specification: {spec}\n"
+            "Try one of:\n"
+            "  PATH.gpkg\n"
+            "  postgresql://HOST/DBNAME[/DBSCHEMA[/TABLE]]\n"
+            "  mssql://HOST/DBNAME[/DBSCHEMA[/TABLE]]\n"
+            "  mysql://HOST[/DBNAME[/TABLE]]"
+        )
+
+    def __init__(
+        self,
+        original_spec,
+        db_type,
+        engine,
+        db_schema,
+        table,
+        **meta_overrides,
+    ):
```

Comment:
with this many arguments, most/all should probably be keyword-only
```suggestion
    def __init__(
        self,
        original_spec,
        *,
        db_type,
        engine,
        db_schema,
        table,
        **meta_overrides,
    ):
```
(note: also requires changes to callers)

Reply:
Done